File: graphify/src/enggraph/features.py
"""Whether a background feature runs for a project, and where that was decided.

A feature lives beside the selection and the schedule, in
`project_settings.settings`, and its fields resolve the way a schedule's do:
the project, then the organizations holding it, then the global default, most
specific first and one field at a time.

One field does not resolve that way. `enabled` at the global level is a gate
rather than a default: turned off there, it is off for every project whatever
their own rows say. That is what makes the global toggle something an operator
can trust - the model can be shut down and the queue stops, without walking
every project first.
"""

from __future__ import annotations

import logging
from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from urllib.parse import urlparse

from psycopg2.extensions import cursor as Cursor

from enggraph.config import (
    FEATURE_DEFAULTS,
    FEATURE_KEY_TTL_DAYS,
    SETTINGS_PROJECT,
)
from enggraph.selection import Origin, levels
from enggraph.storage import read_settings_json

LOG = logging.getLogger(__name__)

# Whether the feature may run at all. Read from the global level and from
# nowhere else: allowing something is not a permission a project grants
# itself. It is deliberately apart from `enabled` - one field doing both jobs
# meant that turning a feature off by default also stopped any project from
# turning itself on, which is the opposite of a default.
ALLOWED = "allowed"
ENABLED = "enabled"
SERVER_URL = "server_url"
# The token a published server wants, and when it was last written. The pair
# travels together: a key with no date could not be reported as due for
# rotation, which is the only thing the date is for.
SERVER_KEY = "server_key"
KEY_SAVED_AT = "key_saved_at"
# How fast the queue behind this feature is worked. `batch` is per project -
# how many files one claim takes, which is how one project is given more of
# the machine than another. The other two are the loop's own and are read
# from the global level: a poll interval and a work budget belong to the
# process, not to whichever project it happens to be draining.
BATCH = "batch"
TICK_SECONDS = "tick_seconds"
BUDGET_SECONDS = "budget_seconds"
# How a file of this project is cut up before it is embedded. A tree of terse
# configuration files and one of long source files are not well served by the
# same window, which is why this is a project's setting rather than the
# stack's. The size is recorded on every chunk, so changing it re-embeds the
# project rather than leaving the old cut in place.
CHUNK_CHARS = "chunk_chars"
CHUNK_OVERLAP = "chunk_overlap"
NUMBERS = (BATCH, TICK_SECONDS, BUDGET_SECONDS, CHUNK_CHARS, CHUNK_OVERLAP)
FIELDS = (ENABLED, SERVER_URL, SERVER_KEY, KEY_SAVED_AT, *NUMBERS)
# Clamped rather than refused when read, as a schedule's minutes are: these
# rows are editable in psql, and a batch of zero would spin a loop.
BOUNDS: dict[str, tuple[int, int]] = {
    BATCH: (1, 64),
    TICK_SECONDS: (1, 3600),
    BUDGET_SECONDS: (5, 3600),
    CHUNK_CHARS: (200, 20000),
    CHUNK_OVERLAP: (0, 200),
}


@dataclass(frozen=True)
class Feature:
    """What a feature does for one project, and where each field came from."""

    name: str
    allowed: bool
    enabled: bool
    server_url: str
    origins: dict[str, Origin]
    # Everything below has a default so that a caller building one by hand -
    # a test, mostly - says only what it cares about, and a field added here
    # does not break every such caller.
    server_key: str = ""
    key_saved_at: str = ""
    batch: int = 1
    tick_seconds: int = 30
    budget_seconds: int = 60
    chunk_chars: int = 1500
    chunk_overlap: int = 5

    @property
    def gated(self) -> bool:
        """Whether the feature is off because it is disabled everywhere.

        Not the same as being off by default: a project may switch itself on
        under that, and under this it is not asked at all.
        """
        return not self.allowed

    @property
    def key_expired(self) -> bool:
        """Whether the stored key is older than the rotation window.

        Advisory, and deliberately so: the key goes on being sent. A queue
        that stopped itself because a date passed would be the silent failure
        this stack keeps being asked to avoid - the dashboard says the key is
        due for renewal, and the server is the one that decides whether it
        still works.
        """
        due = self.key_due()
        return due is not None and due < datetime.now(UTC)

    def key_due(self) -> datetime | None:
        """When the stored key should be renewed, or None when there is none."""
        if not self.server_key or not self.key_saved_at:
            return None
        try:
            saved = datetime.fromisoformat(self.key_saved_at.replace("Z", "+00:00"))
        except ValueError:
            LOG.warning("feature key_saved_at %r is not a date", self.key_saved_at)
            return None
        if saved.tzinfo is None:
            saved = saved.replace(tzinfo=UTC)
        return saved + timedelta(days=FEATURE_KEY_TTL_DAYS)


def clean_enabled(value: object) -> bool | None:
    """Return a stored switch fit to act on, or None when it names none.

    An absent field is not a complaint: it is how a level declines to answer,
    and warning about it filled the log with a line per project per tick.
    """
    if isinstance(value, bool):
        return value
    if value is not None:
        LOG.warning("feature switch is %r, which is not true or false", value)
    return None


def clean_url(value: object) -> str | None:
    """Return a stored server URL fit to dial, or None when it is not one.

    Refused here as well as in the dashboard: these rows are editable in psql,
    and a URL with no scheme reaches urllib as a relative path and fails a
    batch at a time rather than at the point it was typed.
    """
    if not isinstance(value, str):
        LOG.warning("feature server_url is %r, which is not a string", value)
        return None
    url = value.strip().rstrip("/")
    if not url:
        return ""
    parsed = urlparse(url)
    if parsed.scheme not in ("http", "https") or not parsed.netloc:
        LOG.warning("feature server_url %r is not an http or https URL", value)
        return None
    return url


def clean_text(field: str, value: object) -> str | None:
    """Return a stored string fit to use, or None when it is not one."""
    if not isinstance(value, str):
        LOG.warning("feature %s is %r, which is not a string", field, value)
        return None
    return value.strip()


def clean_number(field: str, value: object) -> int | None:
    """Return a stored count clamped into its bounds, or None if it is not one."""
    if isinstance(value, bool) or not isinstance(value, int):
        LOG.warning("feature %s is %r, which is not a whole number", field, value)
        return None
    low, high = BOUNDS[field]
    clamped = max(low, min(high, value))
    if clamped != value:
        LOG.warning("feature %s of %d clamped to %d", field, value, clamped)
    return clamped


def clean(field: str, value: object) -> bool | str | int | None:
    """Return a stored value fit to act on, whichever field it belongs to."""
    if field == ENABLED:
        return clean_enabled(value)
    if field == SERVER_URL:
        return clean_url(value)
    if field in NUMBERS:
        return clean_number(field, value)
    return clean_text(field, value)


def stored(cursor: Cursor, project: str, feature: str) -> dict:
    """Read one level's object for a feature, empty when it says nothing."""
    value = read_settings_json(cursor, project).get(feature)
    return value if isinstance(value, dict) else {}


def resolve(cursor: Cursor, project: str, feature: str) -> Feature:
    """Settle a feature for one project, and say where each field came from.

    `allowed` is read first, from the global level, and answered from at once
    when it says no. Everything else is the schedule's walk: the first level
    that states a field decides it, and a level that states nothing is skipped
    rather than treated as a no - which is what lets a project run a feature
    the global default leaves off.
    """
    defaults = FEATURE_DEFAULTS[feature]
    allowed = clean_enabled(stored(cursor, SETTINGS_PROJECT, feature).get(ALLOWED))
    if allowed is False:
        return Feature(
            name=feature,
            allowed=False,
            enabled=False,
            server_url="",
            server_key="",
            key_saved_at="",
            batch=int(FEATURE_DEFAULTS[feature][BATCH]),
            tick_seconds=int(FEATURE_DEFAULTS[feature][TICK_SECONDS]),
            budget_seconds=int(FEATURE_DEFAULTS[feature][BUDGET_SECONDS]),
            chunk_chars=int(FEATURE_DEFAULTS[feature][CHUNK_CHARS]),
            chunk_overlap=int(FEATURE_DEFAULTS[feature][CHUNK_OVERLAP]),
            origins=dict.fromkeys(FIELDS, "global"),
        )

    found: dict[str, tuple[Origin, bool | str | int]] = {}
    for origin, name in levels(cursor, project):
        level = stored(cursor, name, feature)
        for field in FIELDS:
            if field in found or level.get(field) is None:
                continue
            value = clean(field, level[field])
            if value is not None and value != "":
                found[field] = (origin, value)
    settled = {
        field: found.get(field, ("default", defaults[field])) for field in FIELDS
    }
    return Feature(
        name=feature,
        allowed=True,
        enabled=bool(settled[ENABLED][1]),
        server_url=str(settled[SERVER_URL][1]),
        server_key=str(settled[SERVER_KEY][1]),
        key_saved_at=str(settled[KEY_SAVED_AT][1]),
        batch=int(settled[BATCH][1]),
        tick_seconds=int(settled[TICK_SECONDS][1]),
        budget_seconds=int(settled[BUDGET_SECONDS][1]),
        chunk_chars=int(settled[CHUNK_CHARS][1]),
        chunk_overlap=int(settled[CHUNK_OVERLAP][1]),
        origins={field: settled[field][0] for field in FIELDS},
    )

File: web/src/server/features.ts
import { badRequest, readBodyNumber, readBodyString } from "./args.js";

// The switchable features, as enggraph.config names them. `indexing` is the
// key the schedule already lives under: the switch is another field of that
// object rather than a second place to look.
export const FEATURES = ["indexing", "summarize", "embedding"] as const;
export type FeatureName = (typeof FEATURES)[number];

// A null field is this level letting go of it: stored as a removal, so the
// level above answers. An absent field leaves whatever is stored alone.
export type Feature = {
  // Whether the feature may run at all. Only the global level stores it: a
  // project cannot allow itself something the operator switched off.
  allowed?: boolean;
  enabled?: boolean | null;
  server_url?: string | null;
  server_key?: string;
  key_saved_at?: string;
  // How fast the queue behind this feature is worked. `batch` and the chunk
  // pair are per project; the other two belong to the loop and are global.
  batch?: number | null;
  tick_seconds?: number | null;
  budget_seconds?: number | null;
  chunk_chars?: number | null;
  chunk_overlap?: number | null;
};

// The same bounds enggraph.features clamps to when it reads a row. A value
// refused here can still arrive through psql, so neither side is the only
// guard.
const NUMBERS: Record<string, [number, number]> = {
  batch: [1, 64],
  tick_seconds: [1, 3600],
  budget_seconds: [5, 3600],
  chunk_chars: [200, 20000],
  chunk_overlap: [0, 200],
};
// Only the loop reads these, and the loop reads the global level.
const ROOT_ONLY = new Set(["tick_seconds", "budget_seconds"]);

/** Name a feature, or refuse: the path segment reaches the database as a key. */
export function requireFeature(value: string): FeatureName {
  if (!FEATURES.includes(value as FeatureName)) {
    throw badRequest(
      `Unknown feature "${value}", one of ${FEATURES.join(", ")}`,
    );
  }
  return value as FeatureName;
}

/**
 * Check a server URL before it is stored.
 *
 * Refused here as well as in Python, for the reason the schedule's bounds are
 * checked in both: a row can be written in psql, so neither side is the only
 * guard - but a URL typed in the dashboard should be refused where it was
 * typed rather than by a queue that quietly stops an hour later.
 */
function readServerUrl(body: unknown): string | undefined {
  if ((body as Record<string, unknown>).server_url === null) {
    return "";
  }
  const value = readBodyString(body, "server_url");
  if (value === undefined) {
    return undefined;
  }
  const url = value.trim().replace(/\/$/, "");
  if (url === "") {
    return "";
  }
  let parsed: URL;
  try {
    parsed = new URL(url);
  } catch {
    throw badRequest('Field "server_url" must be an http or https URL');
  }
  if (parsed.protocol !== "http:" && parsed.protocol !== "https:") {
    throw badRequest('Field "server_url" must be an http or https URL');
  }
  return url;
}

/**
 * Read a feature out of a request body, or null when it states nothing.
 *
 * A body that leaves every field out is a level with nothing to say, which is
 * stored by removing the key rather than by writing an empty object - the
 * same rule the schedule follows, and what sends the question back to the
 * level above.
 */
function readSwitch(body: unknown, name: string): boolean | null | undefined {
  const value = (body as Record<string, unknown>)[name];
  if (value === undefined || value === null) {
    return value;
  }
  if (typeof value !== "boolean") {
    throw badRequest(`Field "${name}" must be true or false`);
  }
  return value;
}

function present(body: unknown, name: string): boolean {
  return (body as Record<string, unknown>)[name] !== undefined;
}

export function readFeature(body: unknown, root = false): Feature | null {
  if (body === null || typeof body !== "object") {
    throw badRequest("A JSON object body is required");
  }
  const value: Feature = {};
  // `allowed` is dropped rather than refused below the global level: the
  // dashboard sends one shape, and a project has nothing to say about it.
  const allowed = readSwitch(body, "allowed");
  if (root && typeof allowed === "boolean") {
    value.allowed = allowed;
  }
  const enabled = readSwitch(body, "enabled");
  if (enabled !== undefined) {
    value.enabled = enabled;
  }
  const url = readServerUrl(body);
  if (url !== undefined) {
    value.server_url = url === "" ? null : url;
  }
  for (const [name, [low, high]] of Object.entries(NUMBERS)) {
    // The loop's own pace is only meaningful where the loop reads it, which
    // is the global level; a project sets how big its own claims are.
    if ((!root && ROOT_ONLY.has(name)) || !present(body, name)) {
      continue;
    }
    (value as Record<string, unknown>)[name] =
      readBodyNumber(body, name, low, high) ?? null;
  }
  // Write-only. An absent or empty field leaves whatever is stored alone,
  // which is what makes the input in the dashboard a way to replace a token
  // rather than a way to wipe one by saving the form it is not typed into.
  const key = readBodyString(body, "server_key")?.trim();
  if (key !== undefined && key !== "") {
    value.server_key = key;
    value.key_saved_at = new Date().toISOString();
  }
  return Object.keys(value).length === 0 ? null : value;
}

// How long a stored token is good for before the dashboard asks for a new
// one. The other copy of this number is enggraph.config.FEATURE_KEY_TTL_DAYS,
// which is what the Python side reports; they are read side by side on one
// page, so they have to agree.
const KEY_TTL_DAYS = 30;

type Redacted = Record<string, unknown>;

/**
 * Take the tokens out of a settings object before it is sent to a browser.
 *
 * The dashboard has no authentication of its own: a secret that travelled
 * back to it would be readable by anyone who can open the page, and by
 * anything that caches the response. What a reader needs is whether a token
 * is stored and whether it is due for rotation, and neither is the secret.
 */
export function redactKeys(settings: unknown): Redacted | null {
  if (settings === null || typeof settings !== "object") {
    return null;
  }
  const out: Redacted = {};
  for (const [key, value] of Object.entries(settings as Redacted)) {
    if (value === null || typeof value !== "object") {
      out[key] = value;
      continue;
    }
    const feature = { ...(value as Redacted) };
    const stored = feature.server_key;
    const savedAt = feature.key_saved_at;
    delete feature.server_key;
    if (typeof stored === "string" && stored !== "") {
      const saved = typeof savedAt === "string" ? Date.parse(savedAt) : NaN;
      const due = Number.isNaN(saved)
        ? null
        : new Date(saved + KEY_TTL_DAYS * 86_400_000);
      feature.key_set = true;
      feature.key_due = due === null ? null : due.toISOString();
      feature.key_expired = due !== null && due.getTime() < Date.now();
    }
    out[key] = feature;
  }
  return out;
}
